src/Microsoft.VisualStudio.Threading/SingleThreadedSynchronizationContext.cs

```diff
@@ -0,0 +1,236 @@
+    /// </summary>
+    /// <remarks>
+    /// This must be created on the thread that will serve as the pumping thread.
+    /// </remarks>
+    public class SingleThreadedSynchronizationContext : SynchronizationContext
+    {
+        /// <summary>
+        /// The list of posted messages to be executed. Must be locked for all access.
+        /// </summary>
+        private readonly Queue<Message> messageQueue = new Queue<Message>();
+
+        /// <summary>
+        /// The managed thread ID of the thread this instance owns.
+        /// </summary>
+        private readonly int ownedThreadId = Environment.CurrentManagedThreadId;
+
+        /// <inheritdoc/>
+        public override void Post(SendOrPostCallback d, object state)
+        {
+            var ctxt = ExecutionContext.Capture();
+            lock (this.messageQueue)
+            {
+                this.messageQueue.Enqueue(new Message(d, state, ctxt));
+                Monitor.PulseAll(this.messageQueue);
+            }
+        }
+
+        /// <inheritdoc/>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031", Justification = "We are catching it to rethrow elsewhere.")]
+        public override void Send(SendOrPostCallback d, object state)
+        {
+            Requires.NotNull(d, nameof(d));
+
+            if (this.ownedThreadId == Environment.CurrentManagedThreadId)
+            {
+                try
+                {
+                    d(state);
+                }
+                catch (Exception ex)
+                {
+                    throw new TargetInvocationException(ex);
+                }
+            }
+            else
+            {
+                Exception caughtException = null;
+                var evt = new ManualResetEventSlim();
+                var ctxt = ExecutionContext.Capture();
+                lock (this.messageQueue)
+                {
+                    this.messageQueue.Enqueue(new Message(
+                        s =>
+                        {
+                            try
+                            {
+                                d(state);
+                            }
+                            catch (Exception ex)
+                            {
+                                caughtException = ex;
+                            }
+                            finally
+                            {
+                                evt.Set();
+                            }
+                        },
+                        null,
+                        ctxt));
+                    Monitor.PulseAll(this.messageQueue);
+                }
+
+                evt.Wait();
+                if (caughtException != null)
+                {
+                    throw new TargetInvocationException(caughtException);
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public override SynchronizationContext CreateCopy() => this;
```

Comment:
If I understand it correctly CreateCopy isn't supposed to reuse the instance because reference equality of SynchronizationContext is used to inline invocations in some code paths. This could cause inlining invocations on the wrong thread if SynchronizationContext is flown to a different thread via ExecutionContext. At least thats what I read out of the comments in the WPF SynchronizationContext ([comment on the compat switch here](https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/BaseCompatibilityPreferences.cs,39))

Reply:
Thanks for the comment link. I didn't expect other code to use reference equality to mean they could skip using the SyncContext. Yikes. I'll fix this.